File: sqrt.erl
-module(sqrt).
-export([my_sqrt/1]).

%% x 的平方根
%% https://leetcode-cn.com/problems/sqrtx/description

my_sqrt(X, Low, High) ->
    Mid = (Low + High) / 2,
    M2 = Mid * Mid,
    if
        abs(M2 - X) < 0.00001 -> trunc(Mid);
        M2 < X -> my_sqrt(X, Mid + 1, High);
        true -> my_sqrt(X, Low, Mid - 1)
    end.

-spec my_sqrt(X :: integer()) -> integer().
my_sqrt(X) ->
  if
      X =< 1 -> X;
      true -> my_sqrt(X, 0, X)
  end.

File: sqrt.ex
# x 的平方根
# https://leetcode-cn.com/problems/sqrtx/description

defmodule Solution do
  @spec my_sqrt(x :: integer) :: integer
  def my_sqrt(x) do
    if x <= 1 do
      x
    else
      my_sqrt(x, 0, x)
    end
  end

  defp my_sqrt(x, low, high) do
    mid = (low + high) / 2
    m2 = mid * mid

    cond do
      abs(m2 - x) < 0.00001 -> trunc(mid)
      m2 < x -> my_sqrt(x, mid + 1, high)
      true -> my_sqrt(x, low, mid - 1)
    end
  end
end
